### PR TITLE
PLANNER-932: Made GUI more stable

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**

### CODE REVIEW

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/Calendar.java [hunk 1]
```diff
@@ -101,10 +118,23 @@ public LocalDateTime getViewEndDate() {
     }
 
     public void addShift(I shift) {
-        shifts.add(shift);
+        shifts.put(shift, shift);
         getView().addShift(shift);
     }
 
+    //returns original shift; add it if it doesn't exist
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/Calendar.java]: code style: "// Returns ..."
(space after // and start with capital)

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/TimeSlotTableView.java [hunk 2]
```diff
@@ -160,11 +163,27 @@ public void addTimeSlot(I shift) {
...
         timeSlots.add(drawable);
         updateTimeSlotsFor(shift.getGroupId());
 
     }
 
+    public void updateTimeSlot(I oldShift, I newShift) {
+        if (!oldShift.getGroupId().equals(newShift.getGroupId()) ||
+                !oldShift.getStartTime().equals(newShift.getStartTime()) ||
+                !oldShift.getEndTime().equals(newShift.getEndTime())) {
+            throw new RuntimeException("Old Shift does not exist in the same time slot as New Shift");
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/TimeSlotTableView.java]: Don't throw RuntimeException, that's a too high level class. Use IllegalStateException, IllegalArgumentException, etc

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java [hunk 3]
```diff
@@ -35,88 +35,28 @@
     EmployeeRosterView last;
     Calendar<EmployeeId, EmployeeData> calendar;
     static final String LOADING_STRING = "Fetching Employee Roster";
+    boolean busy;
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java]: buys with what? => imprive variable naming

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java [hunk 4]
```diff
@@ -35,88 +35,28 @@
     EmployeeRosterView last;
     Calendar<EmployeeId, EmployeeData> calendar;
     static final String LOADING_STRING = "Fetching Employee Roster";
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java]: i18n

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java [hunk 5]
```diff
@@ -142,40 +83,97 @@ public void onSuccess(EmployeeRosterView employeeRosterView) {
...
+                                        } else {
+                                            Shift shift = new Shift();
+                                            shift.setTenantId(employee.getTenantId());
+                                            shift.setEmployee(employee);
+                                            shift.setTimeSlot(timeslot);
+                                            EmployeeData newShift = new EmployeeData(shift,
+                                                    timeSlotIdToEmployeeIdToAvailabilityViewMap.get(timeslot.getId())
+                                                            .get(
+                                                                    employee
+                                                                            .getId()));
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java]: code style: unneeded newlines

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java (hunk 4 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeDataFetchable.java]: Actually not i18n; it is a unique identifier String used by LoadingPopup; it isn't display anywhere; probably can be moved to an enum.

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/spot/SpotDataFetchable.java [hunk 6]
```diff
@@ -45,105 +47,98 @@ public SpotDataFetchable(Calendar<SpotId, SpotData> calendar, Provider<Integer>
...
+                            List<TimeSlot> timeslots = spotRosterView.getTimeSlotList();
+                            List<Spot> spots = spotRosterView.getSpotList();
+                            Collection<SpotData> out = new ArrayList<>();
+
+                            for (TimeSlot timeslot : timeslots) {
+                                for (Spot spot : spots) {
+                                    if (null != timeSlotIdToSpotIdToShiftViewListMap.getOrDefault(timeslot.getId(),
+                                            Collections.emptyMap()).get(spot
+                                                    .getId())) {
+                                        timeSlotIdToSpotIdToShiftViewListMap.get(timeslot.getId()).get(spot.getId())
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/spot/SpotDataFetchable.java]: indentation

**Pull request merged**